File: hnsw_pip.py
import numpy as np
import heapq
import math


class HNSW_PiP:
    def __init__(self, dim, M, efConstruction, metric='l2', seed=42,
                 pip_gamma=95.0, pip_delta=20):
        self.dim = dim
        self.M = int(M)
        self.M0 = 2 * int(M)
        self.efConstruction = int(efConstruction)
        self.layers = []
        self.vectors = {}
        self.metric = metric
        self.maxlevel = -1
        self.rng = np.random.default_rng(seed)
        self.entry_id = None
        self.mL = 1.0 / math.log(self.M)
        self.use_heuristic = True

        # PiP paper parameters
        self.pip_gamma = float(pip_gamma)   # γ
        self.pip_delta = int(pip_delta)     # Δ

    """
    From paper algorithm 1: full insertion
               algorithm 2: searching layer to get candidate list
               algorithm 3/4: selecting M neighbors for connection

    PiP modification:
    - construction remains standard HNSW
    - query-time search at layer 0 uses patience-based early termination
    """

    def max_level(self) -> int:
        return self.maxlevel

    def levels(self) -> list:
        return list(range(self.maxlevel + 1))

    def _insert_(self, vector, node_id=None):
        vec = np.asarray(vector, dtype=float)
        if vec.shape[-1] != self.dim:
            raise ValueError(f"Expected dim={self.dim}, got {vec.shape}")

        if node_id is None:
            node_id = len(self.vectors)
        if node_id in self.vectors:
            raise ValueError(f"Node id {node_id} already exists")

        self.vectors[node_id] = vec
        l = self.probab_levels(self.mL)

        # first node
        if self.entry_id is None:
            for _ in range(l + 1):
                self.layers.append({})
            for lc in range(l + 1):
                self.layers[lc][node_id] = set()
            self.entry_id = node_id
            self.maxlevel = l
            return node_id

        old_top = self.maxlevel
        if l > self.maxlevel:
            for _ in range(self.maxlevel + 1, l + 1):
                self.layers.append({})
            self.maxlevel = l

        ep = self.entry_id
        L = old_top

        # greedy descent on upper layers
        for lc in range(L, l, -1):
            ep = self._search_layer_greedy(vec, ep, lc, ef=1)

        # standard HNSW construction on layers min(L,l) ... 0
        for lc in range(min(L, l), -1, -1):
            if node_id not in self.layers[lc]:
                self.layers[lc][node_id] = set()

            W = self._search_layer_standard(vec, ep, lc, self.efConstruction)

            neighbors = self._select_neighbors_heuristic_paper(
                vec, W, layer=lc,
                M=(self.M if lc > 0 else self.M0),
                extend_candidates=True,
                keep_pruned_connections=False
            )

            for nb in neighbors:
                if nb not in self.layers[lc]:
                    self.layers[lc][nb] = set()

                self.layers[lc][node_id].add(nb)
                self.layers[lc][nb].add(node_id)

                Mmax = self.M0 if lc == 0 else self.M
                if len(self.layers[lc][nb]) > Mmax:
                    pruned = self._prune_connections(nb, lc, Mmax)
                    self.layers[lc][nb] = pruned

            if len(W) > 0:
                ep = W[0]

        if l > old_top:
            self.entry_id = node_id

        return node_id

    def _search_layer_greedy(self, vec, curr_entryPointID: int, lc: int, ef: int = 1) -> int:
        best = curr_entryPointID
        best_dist = self.dist(vec, self.vectors[best])

        while True:
            improved = False
            neighbors = self.layers[lc].get(best, set())

            for nb in neighbors:
                d = self.dist(vec, self.vectors[nb])
                if d < best_dist:
                    best_dist = d
                    best = nb
                    improved = True

            if not improved:
                break

        return best

    def _topk_ids_from_heap(self, heap_obj, k: int):
        if k <= 0 or len(heap_obj) == 0:
            return []
        items = [(-neg_d, node_id) for (neg_d, node_id) in heap_obj]
        items.sort(key=lambda x: x[0])  # ascending distance
        return [node_id for (_, node_id) in items[:k]]

    # Standard HNSW / HNSW* layer search
    def _search_layer_standard(self, vec, ep_id: int, layer: int, ef: int, k_thresh: int = None):
        if layer < 0 or layer >= len(self.layers) or len(self.layers[layer]) == 0:
            return []
        if ep_id not in self.vectors:
            return []
        if ef <= 0:
            return []

        visited = set()
        C = []  # min-heap (distance, id)
        W = []  # max-heap via (-distance, id), size <= ef

        use_k_thresh = (k_thresh is not None and k_thresh > 0)
        Kheap = []  # max-heap via (-distance, id), size <= k_thresh

        def threshold():
            if not use_k_thresh:
                return -W[0][0]
            if len(Kheap) < k_thresh:
                return float("inf")
            return -Kheap[0][0]

        def try_add_topk(d, nb):
            if not use_k_thresh:
                return
            if len(Kheap) < k_thresh:
                heapq.heappush(Kheap, (-d, nb))
            else:
                if d < -Kheap[0][0]:
                    heapq.heapreplace(Kheap, (-d, nb))

        dist_ep = self.dist(vec, self.vectors[ep_id])
        visited.add(ep_id)
        heapq.heappush(C, (dist_ep, ep_id))
        heapq.heappush(W, (-dist_ep, ep_id))
        try_add_topk(dist_ep, ep_id)

        while C:
            dist_c, c_id = heapq.heappop(C)

            if dist_c > threshold():
                break

            for nb in self.layers[layer].get(c_id, set()):
                if nb in visited:
                    continue

                visited.add(nb)
                d = self.dist(vec, self.vectors[nb])

                if len(W) < ef or d < threshold():
                    heapq.heappush(C, (d, nb))

                    if len(W) < ef:
                        heapq.heappush(W, (-d, nb))
                    else:
                        if d < -W[0][0]:
                            heapq.heapreplace(W, (-d, nb))

                    try_add_topk(d, nb)

        result = [(-neg_d, node_id) for (neg_d, node_id) in W]
        result.sort(key=lambda x: x[0])
        return [node_id for (_, node_id) in result]

    # PiP layer search: same HNSW traversal, but with patience-based early stop
    def _search_layer_pip(self, vec, ep_id: int, layer: int, ef: int, k: int):
        if layer < 0 or layer >= len(self.layers) or len(self.layers[layer]) == 0:
            return []
        if ep_id not in self.vectors:
            return []
        if ef <= 0:
            return []
        if k <= 0:
            return []

        visited = set()
        C = []   # min-heap of candidates to expand
        W = []   # max-heap via (-distance, id), size <= ef
        Kheap = []  # current top-k result set for PiP saturation tracking

        def threshold():
            return -W[0][0]

        def try_add_topk(d, nb):
            if len(Kheap) < k:
                heapq.heappush(Kheap, (-d, nb))
            else:
                if d < -Kheap[0][0]:
                    heapq.heapreplace(Kheap, (-d, nb))

        dist_ep = self.dist(vec, self.vectors[ep_id])
        visited.add(ep_id)
        heapq.heappush(C, (dist_ep, ep_id))
        heapq.heappush(W, (-dist_ep, ep_id))
        try_add_topk(dist_ep, ep_id)

        prev_topk = None
        saturation_counter = 0

        while C:
            dist_c, c_id = heapq.heappop(C)

            if dist_c > threshold():
                break

            # expand one candidate
            for nb in self.layers[layer].get(c_id, set()):
                if nb in visited:
                    continue

                visited.add(nb)
                d = self.dist(vec, self.vectors[nb])

                if len(W) < ef or d < threshold():
                    heapq.heappush(C, (d, nb))

                    if len(W) < ef:
                        heapq.heappush(W, (-d, nb))
                    else:
                        if d < -W[0][0]:
                            heapq.heapreplace(W, (-d, nb))

                    try_add_topk(d, nb)

            # PiP paper criterion:
            # phi_h,l(q) = 100 * |N_{h-1,l}(q) ∩ N_{h,l}(q)| / k
            # stop if phi_h,l(q) >= gamma for delta consecutive iterations
            curr_topk = self._topk_ids_from_heap(Kheap, k)

            if len(curr_topk) == k:
                curr_topk_set = set(curr_topk)

                if prev_topk is not None:
                    phi_h = 100.0 * len(curr_topk_set.intersection(prev_topk)) / k

                    if phi_h >= self.pip_gamma:
                        saturation_counter += 1
                    else:
                        saturation_counter = 0

                    if saturation_counter >= self.pip_delta:
                        break

                prev_topk = curr_topk_set

        result = [(-neg_d, node_id) for (neg_d, node_id) in W]
        result.sort(key=lambda x: x[0])
        return [node_id for (_, node_id) in result]

    def _select_neighbors_simple(self, vec, candidates, layer: int, Mmax: int):
        unique_candidates = list(dict.fromkeys(candidates))

        if len(unique_candidates) <= Mmax:
            return unique_candidates

        dist_list = []
        for nb in unique_candidates:
            d = self.dist(vec, self.vectors[nb])
            dist_list.append((d, nb))

        dist_list.sort(key=lambda x: x[0])
        selected = [nb for (d, nb) in dist_list[:Mmax]]
        return selected

    def _select_neighbors_heuristic_paper(
        self,
        q_vec,
        candidates,
        layer: int,
        M: int,
        extend_candidates: bool = True,
        keep_pruned_connections: bool = False
    ):
        if M <= 0:
            return []

        W_set = set(candidates)

        if extend_candidates:
            base = list(W_set)
            for e in base:
                if e not in self.layers[layer]:
                    continue
                for adj in self.layers[layer].get(e, set()):
                    W_set.add(adj)

        cand = [(self.dist(q_vec, self.vectors[e]), e) for e in W_set]
        cand.sort(key=lambda x: x[0])

        R = []
        discarded = []

        for d_qe, e in cand:
            good = True
            for r in R:
                if self.dist(self.vectors[e], self.vectors[r]) < d_qe:
                    good = False
                    break

            if good:
                R.append(e)
                if len(R) == M:
                    break
            else:
                discarded.append((d_qe, e))

        if keep_pruned_connections and len(R) < M:
            discarded.sort(key=lambda x: x[0])
            for _, e in discarded:
                if e not in R:
                    R.append(e)
                    if len(R) == M:
                        break

        return R

    def _query(self, q_vec, K, numSearch, use_hnsw_star: bool = False):
        if self.entry_id is None:
            return []

        ep = self.entry_id
        L = self.maxlevel

        # upper layers: standard greedy descent
        for lc in range(L, 0, -1):
            ep = self._search_layer_greedy(q_vec, ep, lc, ef=1)

        # bottom layer: PiP search
        W = self._search_layer_pip(q_vec, ep, 0, numSearch, k=K)

        return W[:K]

    def probab_levels(self, l):
        U = max(self.rng.random(), 1e-12)
        return int(-math.log(U) * l)

    def search(self, Xq: np.ndarray, k: int, efSearch: int):
        Xq = np.asarray(Xq, dtype=np.float32)
        I = np.empty((Xq.shape[0], k), dtype=np.int32)
        D = np.empty((Xq.shape[0], k), dtype=np.float32)

        for i, q in enumerate(Xq):
            ids = self._query(q, K=k, numSearch=efSearch)

            I[i, :len(ids)] = ids
            if len(ids) < k:
                I[i, len(ids):] = -1

            for j in range(k):
                idx = I[i, j]
                if idx == -1:
                    D[i, j] = np.inf
                else:
                    D[i, j] = self.dist(q, self.vectors[int(idx)])

        return D, I

    def dist(self, a: np.ndarray, b: np.ndarray) -> float:
        if self.metric == 'l2':
            diff = a - b
            return float(np.dot(diff, diff))
        elif self.metric == 'cosine':
            denom = (np.linalg.norm(a) * np.linalg.norm(b))
            if denom == 0:
                return 1.0
            return 1.0 - float(np.dot(a, b) / denom)
        else:
            raise ValueError("Unknown metric")

    def entry_point(self):
        return self.entry_id

    def _prune_connections(self, node_id: int, layer: int, Mmax: int):
        neigh_set = self.layers[layer].get(node_id, set())
        if len(neigh_set) <= Mmax:
            return neigh_set

        neighbors = list(neigh_set)
        neighbors = [x for x in neighbors if x != node_id]
        q_vec = self.vectors[node_id]

        if getattr(self, "use_heuristic", False):
            new_neigh_list = self._select_neighbors_heuristic_paper(
                q_vec,
                neighbors,
                layer=layer,
                M=Mmax,
                extend_candidates=True,
                keep_pruned_connections=False
            )
        else:
            new_neigh_list = self._select_neighbors_simple(
                q_vec, neighbors, layer, Mmax
            )

        new_neigh_set = set(new_neigh_list)
        removed = neigh_set - new_neigh_set

        for nb in removed:
            if nb in self.layers[layer]:
                self.layers[layer][nb].discard(node_id)

        self.layers[layer][node_id] = new_neigh_set
        return new_neigh_set

File: hnsw_pip_cpp/src/hnsw_pip.cpp
#include "hnsw_pip.h"

HNSWPiPIndex::HNSWPiPIndex(int dim,
                           int M,
                           int efConstruction,
                           const std::string& metric,
                           uint32_t seed,
                           double pip_gamma,
                           int pip_delta)
    : dim_(dim),
      M_(M),
      M0_(2 * M),
      efConstruction_(efConstruction),
      metric_(metric),
      maxlevel_(-1),
      entry_id_(-1),
      mL_(1.0 / std::log((double)M)),
      use_heuristic_(true),
      pip_gamma_(pip_gamma),
      pip_delta_(pip_delta),
      rng_(seed),
      uni_(0.0, 1.0) {
    if (dim_ <= 0) throw std::invalid_argument("dim must be positive");
    if (M_ <= 1) throw std::invalid_argument("M must be > 1");
    if (metric_ != "l2" && metric_ != "cosine") {
        throw std::invalid_argument("metric must be 'l2' or 'cosine'");
    }
}

float HNSWPiPIndex::dist(const std::vector<float>& a, const std::vector<float>& b) const {
    if ((int)a.size() != dim_ || (int)b.size() != dim_) {
        throw std::runtime_error("dimension mismatch in dist()");
    }

    if (metric_ == "l2") {
        float s = 0.0f;
        for (int i = 0; i < dim_; ++i) {
            float d = a[i] - b[i];
            s += d * d;
        }
        return s;
    }

    double dot = 0.0, na = 0.0, nb = 0.0;
    for (int i = 0; i < dim_; ++i) {
        dot += (double)a[i] * b[i];
        na += (double)a[i] * a[i];
        nb += (double)b[i] * b[i];
    }
    double denom = std::sqrt(na) * std::sqrt(nb);
    if (denom == 0.0) return 1.0f;
    return (float)(1.0 - dot / denom);
}

int HNSWPiPIndex::sample_level() {
    double U = std::max(uni_(rng_), 1e-12);
    return (int)(-std::log(U) * mL_);
}

void HNSWPiPIndex::add(const std::vector<std::vector<float>>& X) {
    for (int i = 0; i < (int)X.size(); ++i) {
        add_point(X[i], (int)vectors_.size());
    }
}

int HNSWPiPIndex::add_point(const std::vector<float>& vec, int node_id) {
    if ((int)vec.size() != dim_) {
        throw std::invalid_argument("vector dimension mismatch");
    }

    if (node_id < 0) node_id = (int)vectors_.size();
    if (node_id != (int)vectors_.size()) {
        throw std::invalid_argument("this standalone version expects sequential node ids");
    }

    vectors_.push_back(vec);
    int l = sample_level();

    if (entry_id_ == -1) {
        layers_.resize(l + 1);
        for (int lc = 0; lc <= l; ++lc) {
            layers_[lc].resize(node_id + 1);
            layers_[lc][node_id] = {};
        }
        entry_id_ = node_id;
        maxlevel_ = l;
        return node_id;
    }

    int old_top = maxlevel_;
    if (l > maxlevel_) {
        layers_.resize(l + 1);
        maxlevel_ = l;
    }

    for (auto& layer : layers_) {
        if ((int)layer.size() <= node_id) {
            layer.resize(node_id + 1);
        }
    }

    int ep = entry_id_;
    int L = old_top;

    for (int lc = L; lc > l; --lc) {
        ep = search_layer_greedy(vec, ep, lc);
    }

    for (int lc = std::min(L, l); lc >= 0; --lc) {
        std::vector<int> W = search_layer_standard(vec, ep, lc, efConstruction_);

        std::vector<int> neighbors = select_neighbors_heuristic(
            vec, W, lc, (lc > 0 ? M_ : M0_), true, false
        );

        for (int nb : neighbors) {
            layers_[lc][node_id].insert(nb);
            layers_[lc][nb].insert(node_id);

            int Mmax = (lc == 0 ? M0_ : M_);
            if ((int)layers_[lc][nb].size() > Mmax) {
                layers_[lc][nb] = prune_connections(nb, lc, Mmax);
            }
        }

        if (!W.empty()) ep = W[0];
    }

    if (l > old_top) {
        entry_id_ = node_id;
    }

    return node_id;
}

int HNSWPiPIndex::search_layer_greedy(const std::vector<float>& q, int ep, int lc) const {
    int best = ep;
    float best_dist = dist(q, vectors_[best]);

    while (true) {
        bool improved = false;
        for (int nb : layers_[lc][best]) {
            float d = dist(q, vectors_[nb]);
            if (d < best_dist) {
                best_dist = d;
                best = nb;
                improved = true;
            }
        }
        if (!improved) break;
    }
    return best;
}

std::vector<int> HNSWPiPIndex::search_layer_standard(const std::vector<float>& q,
                                                     int ep,
                                                     int layer,
                                                     int ef) const {
    if (layer < 0 || layer >= (int)layers_.size()) return {};
    if (ep < 0 || ep >= (int)vectors_.size()) return {};
    if (ef <= 0) return {};

    std::unordered_set<int> visited;
    std::priority_queue<CandidateMin, std::vector<CandidateMin>, std::greater<CandidateMin>> C;
    std::priority_queue<CandidateMax> W;

    float dist_ep = dist(q, vectors_[ep]);
    visited.insert(ep);
    C.push({dist_ep, ep});
    W.push({dist_ep, ep});

    while (!C.empty()) {
        CandidateMin cur = C.top();
        C.pop();

        float worst_dist = W.top().dist;
        if (cur.dist > worst_dist) break;

        for (int nb : layers_[layer][cur.id]) {
            if (visited.find(nb) != visited.end()) continue;
            visited.insert(nb);

            float d = dist(q, vectors_[nb]);

            if ((int)W.size() < ef) {
                C.push({d, nb});
                W.push({d, nb});
            } else {
                worst_dist = W.top().dist;
                if (d < worst_dist) {
                    C.push({d, nb});
                    W.pop();
                    W.push({d, nb});
                }
            }
        }
    }

    std::vector<std::pair<float, int>> result;
    while (!W.empty()) {
        result.push_back({W.top().dist, W.top().id});
        W.pop();
    }
    std::sort(result.begin(), result.end(),
              [](const auto& a, const auto& b) { return a.first < b.first; });

    std::vector<int> ids;
    ids.reserve(result.size());
    for (const auto& p : result) ids.push_back(p.second);
    return ids;
}

std::vector<int> HNSWPiPIndex::heap_topk_ids(const std::priority_queue<CandidateMax>& heap, int k) {
    auto copy = heap;
    std::vector<std::pair<float, int>> items;
    while (!copy.empty()) {
        items.push_back({copy.top().dist, copy.top().id});
        copy.pop();
    }
    std::sort(items.begin(), items.end(),
              [](const auto& a, const auto& b) { return a.first < b.first; });

    std::vector<int> ids;
    for (int i = 0; i < std::min(k, (int)items.size()); ++i) {
        ids.push_back(items[i].second);
    }
    return ids;
}

std::vector<int> HNSWPiPIndex::search_layer_pip(const std::vector<float>& q,
                                                int ep,
                                                int layer,
                                                int ef,
                                                int k) const {
    if (layer < 0 || layer >= (int)layers_.size()) return {};
    if (ep < 0 || ep >= (int)vectors_.size()) return {};
    if (ef <= 0 || k <= 0) return {};

    std::unordered_set<int> visited;
    std::priority_queue<CandidateMin, std::vector<CandidateMin>, std::greater<CandidateMin>> C;
    std::priority_queue<CandidateMax> W;
    std::priority_queue<CandidateMax> Kheap;

    auto threshold = [&]() -> float {
        return W.top().dist;
    };

    auto try_add_topk = [&](float d, int id) {
        if ((int)Kheap.size() < k) {
            Kheap.push({d, id});
        } else if (d < Kheap.top().dist) {
            Kheap.pop();
            Kheap.push({d, id});
        }
    };

    float dist_ep = dist(q, vectors_[ep]);
    visited.insert(ep);
    C.push({dist_ep, ep});
    W.push({dist_ep, ep});
    try_add_topk(dist_ep, ep);

    std::unordered_set<int> prev_topk;
    int saturation_counter = 0;

    while (!C.empty()) {
        CandidateMin cur = C.top();
        C.pop();

        if (cur.dist > threshold()) break;

        for (int nb : layers_[layer][cur.id]) {
            if (visited.find(nb) != visited.end()) continue;
            visited.insert(nb);

            float d = dist(q, vectors_[nb]);
            if ((int)W.size() < ef) {
                C.push({d, nb});
                W.push({d, nb});
                try_add_topk(d, nb);
            } else if (d < threshold()) {
                C.push({d, nb});
                W.pop();
                W.push({d, nb});
                try_add_topk(d, nb);
            }
        }

        std::vector<int> curr = heap_topk_ids(Kheap, k);
        if ((int)curr.size() == k) {
            std::unordered_set<int> curr_set(curr.begin(), curr.end());
            if (!prev_topk.empty()) {
                int overlap = 0;
                for (int x : curr_set) {
                    if (prev_topk.find(x) != prev_topk.end()) ++overlap;
                }
                double phi = 100.0 * (double)overlap / (double)k;
                if (phi >= pip_gamma_) ++saturation_counter;
                else saturation_counter = 0;

                if (saturation_counter >= pip_delta_) break;
            }
            prev_topk = std::move(curr_set);
        }
    }

    std::vector<std::pair<float, int>> result;
    while (!W.empty()) {
        result.push_back({W.top().dist, W.top().id});
        W.pop();
    }
    std::sort(result.begin(), result.end(),
              [](const auto& a, const auto& b) { return a.first < b.first; });

    std::vector<int> ids;
    ids.reserve(result.size());
    for (const auto& p : result) ids.push_back(p.second);
    return ids;
}

std::vector<int> HNSWPiPIndex::select_neighbors_simple(const std::vector<float>& q,
                                                       const std::vector<int>& candidates,
                                                       int Mmax) const {
    std::vector<int> unique;
    std::unordered_set<int> seen;
    for (int c : candidates) {
        if (seen.insert(c).second) unique.push_back(c);
    }

    if ((int)unique.size() <= Mmax) return unique;

    std::vector<std::pair<float, int>> ds;
    ds.reserve(unique.size());
    for (int nb : unique) ds.push_back({dist(q, vectors_[nb]), nb});
    std::sort(ds.begin(), ds.end(),
              [](const auto& a, const auto& b) { return a.first < b.first; });

    std::vector<int> out;
    for (int i = 0; i < Mmax; ++i) out.push_back(ds[i].second);
    return out;
}

std::vector<int> HNSWPiPIndex::select_neighbors_heuristic(const std::vector<float>& q,
                                                          const std::vector<int>& candidates,
                                                          int layer,
                                                          int Mmax,
                                                          bool extend_candidates,
                                                          bool keep_pruned_connections) const {
    if (Mmax <= 0) return {};

    std::unordered_set<int> Wset(candidates.begin(), candidates.end());
    if (extend_candidates) {
        std::vector<int> base(Wset.begin(), Wset.end());
        for (int e : base) {
            for (int adj : layers_[layer][e]) {
                Wset.insert(adj);
            }
        }
    }

    std::vector<std::pair<float, int>> cand;
    cand.reserve(Wset.size());
    for (int e : Wset) cand.push_back({dist(q, vectors_[e]), e});
    std::sort(cand.begin(), cand.end(),
              [](const auto& a, const auto& b) { return a.first < b.first; });

    std::vector<int> R;
    std::vector<std::pair<float, int>> discarded;

    for (const auto& it : cand) {
        float d_qe = it.first;
        int e = it.second;
        bool good = true;
        for (int r : R) {
            if (dist(vectors_[e], vectors_[r]) < d_qe) {
                good = false;
                break;
            }
        }
        if (good) {
            R.push_back(e);
            if ((int)R.size() == Mmax) break;
        } else {
            discarded.push_back({d_qe, e});
        }
    }

    if (keep_pruned_connections && (int)R.size() < Mmax) {
        std::sort(discarded.begin(), discarded.end(),
                  [](const auto& a, const auto& b) { return a.first < b.first; });
        for (const auto& it : discarded) {
            if (std::find(R.begin(), R.end(), it.second) == R.end()) {
                R.push_back(it.second);
                if ((int)R.size() == Mmax) break;
            }
        }
    }

    return R;
}

std::unordered_set<int> HNSWPiPIndex::prune_connections(int node_id, int layer, int Mmax) {
    const auto& neigh = layers_[layer][node_id];
    if ((int)neigh.size() <= Mmax) return neigh;

    std::vector<int> neighbors;
    for (int x : neigh) {
        if (x != node_id) neighbors.push_back(x);
    }

    std::vector<int> new_neigh_list =
        use_heuristic_
            ? select_neighbors_heuristic(vectors_[node_id], neighbors, layer, Mmax, true, false)
            : select_neighbors_simple(vectors_[node_id], neighbors, Mmax);

    std::unordered_set<int> new_set(new_neigh_list.begin(), new_neigh_list.end());

    for (int nb : neigh) {
        if (new_set.find(nb) == new_set.end()) {
            layers_[layer][nb].erase(node_id);
        }
    }

    layers_[layer][node_id] = new_set;
    return new_set;
}

std::vector<int> HNSWPiPIndex::query_pip(const std::vector<float>& q, int K, int efSearch) const {
    if (entry_id_ == -1) return {};
    int ep = entry_id_;
    for (int lc = maxlevel_; lc > 0; --lc) {
        ep = search_layer_greedy(q, ep, lc);
    }
    std::vector<int> W = search_layer_pip(q, ep, 0, efSearch, K);
    if ((int)W.size() > K) W.resize(K);
    return W;
}

std::pair<std::vector<std::vector<float>>, std::vector<std::vector<int>>>
HNSWPiPIndex::search(const std::vector<std::vector<float>>& Xq, int k, int efSearch) const {
    std::vector<std::vector<float>> D(Xq.size(), std::vector<float>(k, std::numeric_limits<float>::infinity()));
    std::vector<std::vector<int>> I(Xq.size(), std::vector<int>(k, -1));

    for (size_t i = 0; i < Xq.size(); ++i) {
        std::vector<int> ids = query_pip(Xq[i], k, efSearch);
        for (int j = 0; j < (int)ids.size() && j < k; ++j) {
            I[i][j] = ids[j];
            D[i][j] = dist(Xq[i], vectors_[ids[j]]);
        }
    }
    return {D, I};
}